### DISCUSSION TITLE
Option Troubles  - Most Cases Yield a `Null` Option Value

### DISCUSSION THREAD

**Discussion by rsteppac** [spring-projects/spring-shell#491] [Q&A]
I am brand new to Spring Shell, so it might well be a simple newbie issue... but I am struggling to get command options to work.

Using annotations I can get them to work for simple cases as long as I do not specify a longname (have not tried shortnames). This works, in a way:

```java
    @ShellMethod(value = "Option test 1.", key = "t1")
    public String test1(@ShellOption(help = "A shell option") String myParam) {
        return String.format("T1: You passed in: '%s'", myParam);
    }
```
- `t1 hello` prints `T1: You passed in: 'hello'` - unexpected, but OK?
- `t1 --myParam hello` prints `T1: You passed in: 'hello'` - works as expected
- `t1 --userIsMistaken hello` prints `T1: You passed in: '--userIsMistaken'` - not OK, expected error due to unknown named option

As soon as I add a longname to the option parameter like so

```java
@ShellMethod(value = "Option test 2.", key = "t2")
    public String test2(@ShellOption(help = "A shell option", value = {"alongname"}) String myParam) {
        return String.format("T2: You passed in: '%s'", myParam);
    }
```
I cannot get it to accept the option value at all. All call flavors of `t1` executed for `t2` plus
- `t2 --alongname hello` print `T2: You passed in: 'null'`

Calling `help t2` also does not list the option:

```
shell:>help t2
NAME
       t2 - Option test 2.

SYNOPSIS
       t2 

OPTIONS
```

Using `CommandRegistration` and calling a POJO method I have not managed to communicate an option value at all. E.g. 

```java
    @Bean
    CommandRegistration commandRegistrationT2() {
        //@formatter:off
        return CommandRegistration.builder()
            .command("crT2")
            .group("CR Test Commands")
            .description("CommandRegistration option test 2")
            .withOption()
                .description("A shell option")
                .type(String.class)
                .and()
            .withTarget()
                .method(pojoTestCommands, "echo", String.class)
                .and()
            .build();
        //@formatter:oon
    }
```

with the `echo` method defined like so in a POJO (no shell annotations present at all) 

```java
    public String echo(String myParam) {
        return String.format("You passed in: '%s'", myParam);
    }
```

always yields `You passed in: 'null'`.

Can someone point out where I am going wrong here?

Spring Shell version `2.1.0`
Spring Boot version `2.7.2`
Java version `11`

I am using the plain vanilla project as it is generated by start.spring.io with `Spring Shell` as the only dependency.

Thanks!
Ralf

**Comment by jvalkeal** [comment]
Looks like you're missing argument name like in https://github.com/spring-projects/spring-shell/blob/1ff802d8e044ac01aaeb0f9990020f9c704b5d04/spring-shell-samples/src/main/java/org/springframework/shell/samples/e2e/DefaultValueCommands.java#L45

**Reply by rsteppac** [reply]
That makes no difference. If I provide a longname, set the arity, and make the option mandatory, I still receive a `null` option value in the method the command is configured to call.

The command registration bean:

```Java
    @Bean
    CommandRegistration commandRegistrationT1() {
        //@formatter:off
        return CommandRegistration.builder()
            .command("crT1")
            .group("CR Test Commands")
            .description("CommandRegistration option test 1")
            .withOption()
                .description("A shell option")
                .type(String.class)
                .arity(EXACTLY_ONE)
                .longNames("alongname")
                .required()
            .and()
            .withTarget()
                .method(pojoTestCommands, "echo", String.class)
                .and()
            .build();
        //@formatter:oon
    }
```

The POJO containing the method the command calls:

```java
@Component
public class PojoTestCommands {

    public String echo(String myParam) {
        return String.format("You passed in: '%s'", myParam);
    }

}
```

The shell output:

```
shell:>help crT1
NAME
       crT1 - CommandRegistration option test 1

SYNOPSIS
       crT1 [--alongname String]

OPTIONS
       --alongname String
       A shell option
       [Mandatory]

shell:>crT1 --alongname foo
You passed in: 'null'
shell:>crT1 --alongname "foo"
You passed in: 'null'
shell:>crT1 foo
Missing option, longnames='alongname', shortnames=''
shell:>
```
